packages/react-router-dom/index.tsx

```diff
@@ -1551,6 +1551,12 @@ export function useFetcher<TData = any>({
     setFetcherKey(getUniqueFetcherId());
   }
 
+  React.useEffect(() => {
+    if (key && key !== "" && key != fetcherKey) {
+      setFetcherKey(key);
+    }
+  }, [key, fetcherKey]);
```

Comment:
Can't we set the state directly in render?

Reply:
I'll test that out - I always worry about that based on the "rules of hooks" and what not - but a re-reading of the docs makes it sound like this is a valid use case: https://react.dev/reference/react/useState#storing-information-from-previous-renders